Mathlib/Data/Polynomial/RingDivision.lean

```diff
@@ -394,15 +394,131 @@ theorem pow_rootMultiplicity_not_dvd {p : R[X]} (p0 : p ≠ 0) (a : R) :
     ¬(X - C a) ^ (rootMultiplicity a p + 1) ∣ p := by rw [← rootMultiplicity_le_iff p0]
 #align polynomial.pow_root_multiplicity_not_dvd Polynomial.pow_rootMultiplicity_not_dvd
 
+theorem X_sub_C_pow_dvd_iff {p : R[X]} {t : R} {n : ℕ} :
+    (X - C t) ^ n ∣ p ↔ X ^ n ∣ p.comp (X + C t) := by
+  convert (map_dvd_iff <| algEquivAevalXAddC t).symm using 2
+  simp [C_eq_algebraMap]
+
+theorem comp_X_add_C_eq_zero_iff {p : R[X]} (t : R) :
+    p.comp (X + C t) = 0 ↔ p = 0 := AddEquivClass.map_eq_zero_iff (algEquivAevalXAddC t)
+
+theorem comp_X_add_C_ne_zero_iff {p : R[X]} (t : R) :
+    p.comp (X + C t) ≠ 0 ↔ p ≠ 0 := Iff.not <| comp_X_add_C_eq_zero_iff t
+
+theorem rootMultiplicity_eq_rootMultiplicity {p : R[X]} {t : R} :
+    p.rootMultiplicity t = (p.comp (X + C t)).rootMultiplicity 0 := by
+  classical
+  simp_rw [rootMultiplicity_eq_multiplicity, comp_X_add_C_eq_zero_iff]
+  congr; ext; congr 1
+  rw [C_0, sub_zero]
+  convert (multiplicity.multiplicity_map_eq <| algEquivAevalXAddC t).symm using 2
+  simp [C_eq_algebraMap]
+
+theorem rootMultiplicity_eq_natTrailingDegree' {p : R[X]} :
+    p.rootMultiplicity 0 = p.natTrailingDegree := by
+  by_cases h : p = 0
+  · simp only [h, rootMultiplicity_zero, natTrailingDegree_zero]
+  refine le_antisymm ?_ ?_
+  · rw [rootMultiplicity_le_iff h, map_zero, sub_zero, X_pow_dvd_iff, not_forall]
+    exact ⟨p.natTrailingDegree,
+      fun h' ↦ trailingCoeff_nonzero_iff_nonzero.2 h <| h' <| Nat.lt.base _⟩
+  · rw [le_rootMultiplicity_iff h, map_zero, sub_zero, X_pow_dvd_iff]
+    exact fun _ ↦ coeff_eq_zero_of_lt_natTrailingDegree
+
+theorem rootMultiplicity_eq_natTrailingDegree {p : R[X]} {t : R} :
+    p.rootMultiplicity t = (p.comp (X + C t)).natTrailingDegree :=
+  rootMultiplicity_eq_rootMultiplicity.trans rootMultiplicity_eq_natTrailingDegree'
+
+section nonZeroDivisors
+
+open scoped nonZeroDivisors
+
+theorem Monic.mem_nonZeroDivisors {p : R[X]} (h : p.Monic) : p ∈ R[X]⁰ :=
+  mem_nonZeroDivisors_iff.2 fun _ hx ↦ (mul_left_eq_zero_iff h).1 hx
+
+theorem mem_nonZeroDivisors_of_leadingCoeff {p : R[X]} (h : p.leadingCoeff ∈ R⁰) : p ∈ R[X]⁰ := by
+  refine mem_nonZeroDivisors_iff.2 fun x hx ↦ leadingCoeff_eq_zero.1 ?_
+  by_contra hx'
+  rw [← mul_right_mem_nonZeroDivisors_eq_zero_iff h] at hx'
+  simp only [← leadingCoeff_mul' hx', hx, leadingCoeff_zero, not_true] at hx'
+
+end nonZeroDivisors
+
+theorem rootMultiplicity_mul_X_sub_C_pow {p : R[X]} {a : R} {n : ℕ} (h : p ≠ 0) :
+    (p * (X - C a) ^ n).rootMultiplicity a = p.rootMultiplicity a + n := by
+  have h2 := monic_X_sub_C a |>.pow n |>.mul_left_ne_zero h
+  refine le_antisymm ?_ ?_
+  · rw [rootMultiplicity_le_iff h2]
+    intro ⟨q, h'⟩
+    rw [mul_comm p _, add_comm _ n, add_assoc, pow_add, mul_assoc,
+      mul_cancel_left_mem_nonZeroDivisors (monic_X_sub_C a |>.pow n |>.mem_nonZeroDivisors)] at h'
```

Comment:
Can you please extract `dvd_cancel_left/right_mem_nonZeroDivisors`?

Reply:
There is [mul_dvd_mul_iff_left](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/GroupWithZero/Divisibility.html#mul_dvd_mul_iff_left) and [mul_dvd_mul_iff_right](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/GroupWithZero/Divisibility.html#mul_dvd_mul_iff_right). New lemmas should use different names...